src/dbnode/storage/index/aggregate_results.go

```diff
@@ -129,195 +204,124 @@ func (r *aggregatedResults) Reset(
-			// the underlying pool.
-			f.Finalize()
-		}
-		valuesMap := aggValues.Map()
-		for _, t := range entry.Terms {
-			if !valuesMap.Contains(t) {
-				// we can avoid the copy because we assume ownership of the passed ident.ID,
-				// but still need to finalize it.
-				valuesMap.SetUnsafe(t, struct{}{}, AggregateValuesMapSetUnsafeOptions{
-					NoCopyKey:     true,
-					NoFinalizeKey: false,
-				})
-				valueInsertions++
-			} else {
-				// because we already have a entry for this term, we release the ident back to
-				// the underlying pool.
-				t.Finalize()
-			}
-		}
-	}
-	size := r.resultsMap.Len()
-	docsCount := r.totalDocsCount + valueInsertions
-	r.totalDocsCount = docsCount
-	r.Unlock()
-	return size, docsCount
-}
-
-func (r *aggregatedResults) addDocumentsBatchWithLock(
-	batch []doc.Document,
-) error {
-	for _, doc := range batch { //nolint:gocritic
-		switch r.aggregateOpts.Type {
-		case AggregateTagNamesAndValues:
-			if err := r.addDocumentWithLock(doc); err != nil {
-				return err
-			}
-
-		case AggregateTagNames:
-			// NB: if aggregating by name only, can ignore any additional documents
-			// after the result map size exceeds the optional size limit, since all
-			// incoming terms are either duplicates or new values which will exceed
-			// the limit.
-			size := r.resultsMap.Len()
-			if r.aggregateOpts.SizeLimit > 0 && size >= r.aggregateOpts.SizeLimit {
-				return nil
-			}
-
-			if err := r.addDocumentTermsWithLock(doc); err != nil {
-				return err
-			}
-		default:
-			return fmt.Errorf("unsupported aggregation type: %v", r.aggregateOpts.Type)
-		}
-	}
-
-	return nil
-}
+	defer r.Unlock()
 
-func (r *aggregatedResults) addDocumentTermsWithLock(
-	container doc.Document,
-) error {
-	document, err := docs.MetadataFromDocument(container, &r.encodedDocReader)
-	if err != nil {
-		return fmt.Errorf("unable to decode encoded document; %w", err)
+	// NB: init total count with batch length, since each aggregated entry
+	// will have one field.
+	totalCount := len(batch)
+	for idx := 0; idx < len(batch); idx++ {
+		totalCount += len(batch[idx].Terms)
 	}
-	for _, field := range document.Fields { //nolint:gocritic
-		if err := r.addTermWithLock(field.Name); err != nil {
-			return fmt.Errorf("unable to add document terms [%+v]: %w", document, err)
-		}
-	}
-
-	return nil
-}
 
-func (r *aggregatedResults) addTermWithLock(
-	term []byte,
-) error {
-	if len(term) == 0 {
-		return fmt.Errorf(missingDocumentFields, "term")
+	r.aggregateOpts.AggregateUsageMetrics.IncTotal(int64(totalCount))
+	remainingDocs := math.MaxInt64
+	if r.aggregateOpts.DocsLimit != 0 {
+		remainingDocs = r.aggregateOpts.DocsLimit - r.totalDocsCount
 	}
 
-	// if a term filter is provided, ensure this field matches the filter,
-	// otherwise ignore it.
-	filter := r.aggregateOpts.FieldFilter
-	if filter != nil && !filter.Allow(term) {
-		return nil
-	}
+	// NB: already hit doc limit.
+	if remainingDocs <= 0 {
```

Comment:
Hm for the `Require-Exhaustive: false` case shouldn't we at least up to the limit before failing?

Reply:
Ah I see that we do, given that we're minusing docs limit and what we already have counted as the total docs count, nevermind.